Gems/PhysX/Code/Tests/CharacterGameplayControllerTests.cpp

```diff
@@ -124,4 +133,33 @@ namespace PhysX
             EXPECT_THAT(expectedGravityMultiplier + i, testing::FloatEq(basis.m_gameplayController->GetGravityMultiplier()));
         }
     }
+
+    TEST_F(PhysXDefaultWorldTest, CharacterGameplayController_EntityFallsUnderGravity)
+    {
+        GameplayTestBasis basis(m_testSceneHandle, DefaultGravityMultiplier, DefaultGroundDetectionBoxHeight, DefaultFloorTransform);
+        auto* sceneInterface = AZ::Interface<AzPhysics::SceneInterface>::Get();
+        sceneInterface->RemoveSimulatedBody(basis.m_sceneHandle, basis.m_floor->m_bodyHandle);
+        
+        // Let scene run for a few moments so the entity can be manipulated by gravity from the gameplay component
+        auto startPosition = basis.m_controller->GetPosition();
+        
+        int timeStepCount = 10;
+        float totalTime = 0.0f;
+        float timeStep = AzPhysics::SystemConfiguration::DefaultFixedTimestep;
+        float distanceFell = 0.0f;
+
+        for (int i = 0; i < timeStepCount; i++)
+        {
+            basis.Update();
+            totalTime += timeStep;
+        }
+
+        auto endPosition = basis.m_controller->GetPosition();
+        
+        // calculate distance fallen (d = 0.5 * g * t^2)
+        distanceFell = 0.5f * (-10) * (totalTime * totalTime);
+        
+        EXPECT_THAT(endPosition.GetZ(), testing::FloatNear(startPosition.GetZ() + distanceFell, 0.1f));
```

Comment:
nit: `0.1f` tolerance seems high, might be worth decreasing this to a much smaller value for this sort of test

Reply:
+1

Reply:
@hultonha @moraaar @greerdv 

If I change the tolerance to `0.01f` or lower these test will fail (I've pushed up the code using this value atm).

```
E:/gws/o3de/Gems/PhysX/Code/Tests/CharacterGameplayControllerTests.cpp(170): error: Value of: endPosition.GetZ()
Expected: is approximately -0.1362505 (absolute error <= 0.0099999998)
  Actual: -0.122625 (of type float), which is 0.013625 from -0.13625
```

Is there something I can do to get the calculations to be closer to use a tighter tolerance, or is it ok in this instance to use a tolerance of `0.1f`?

Reply:
Unless the math done to calculate `distanceFell` matches the math used internally by PhysX I'm not sure you'll ever get a reliable comparison. Maybe a more reliable test would be to check the end position is lower than the start position. Tagging @greerdv

Reply:
Yeah I agree, I'm not sure how helpful this is if the tolerance is large and the values don't actually really match

Reply:
The expected value calculation here is what you would get if the simulation was perfect. The discrepancy is coming about because the simulation is stepping time in discrete intervals. When I implemented the character gameplay component I used a pretty naive integration scheme, which takes a snapshot of the velocity at a timestep and applies that over a whole timestep. It looks kind of like this

![image](https://user-images.githubusercontent.com/82226198/219501702-8babb5de-76fb-4ffc-83f8-292f71ae8efa.png)

The upshot of that is that there is a discrepancy in the distance travelled which is roughly equal to half a time step's worth of motion. Because the distance fallen is quadratic, and the total time here is pretty small, that difference is large compared to the distance fallen. It works out that the relative error is about equal to the timestep divided by the total time, so for this example, where the total time is 1/6s and the timestep is 1/60s, you get a relative error of (1/60)/(1/6) = 1/10 so about 10% error. If you simulated for longer e.g. 2s, you would get a relative error (1/60)/2 = 1/120. In that case the distance fallen would be a bit less than 20m, and the tolerance would need to be about 0.2m (about 1%).

I'd still much rather have some idea of how far it should have fallen feed into the expectation. If you just expect it to be lower than it started, you could pass the test whether you'd fallen 1mm or 1km.

Reply:
Based on @greerdv's feedback, I've adjusted the calculations to compensate for the discrete intervals vs continuous and was able to set a much tighter tolerance. 

This also inspired me to use use a bit of variability in the number of time steps taken to test some minor bounds of time that are reasonable for a unit scale test.